### DISCUSSION TITLE
Support Jinja2's ChainableUndefined to reduce attribute errors

### DISCUSSION THREAD

**Discussion by hackery** [saltstack/salt#67353] [Feature Requests]
A common need in Jinja rendered states and files is to access nested attributes of pillar data, but you have to take steps to avoid referencing any undefined values in the chain with many non-trivial merged pillars - using `myapp.cfg.pipeline.credentials.msk.keystore` needs multiple levels of guard to avoid `'dict object' has no attribute 'msk'` type errors. The `default(...)` construct doesn't help except when used for top-level items.

**Describe the solution you'd like**
Any nested element in chained attribute access which evaluates as `Undefined` should not throw an error, but continue to evaluate the whole reference as `Undefined`.

Jinja has [supported](https://jinja.palletsprojects.com/en/stable/templates/#jinja-filters.default) a `ChainableUndefined` environment option since v2.11 in early 2020 which provides exactly this functionality at minimal implementation expense for the Salt project. The one significant complexity I foresee is how this should be *optionally* enabled - whether it can only be a global option, or should have a way to localise it to target states/files (and whether this applies transitively to include/imported files, rendered managed files etc).

**Describe alternatives you've considered**
* Multiple levels of guard on all references - makes files much less readable/maintainable
* Exhaustive setting of default values everywhere in pillar data - not always appropriate, bloats data and reduces maintainability

**Comment by welcome** [comment]
Hi there! Welcome to the Salt Community! Thank you for making your first contribution. We have a lengthy process for issues and PRs. Someone from the Core Team will follow up as soon as possible. In the meantime, here’s some information that may help as you continue your Salt journey.
Please be sure to review our [Code of Conduct](https://github.com/saltstack/salt/blob/master/CODE_OF_CONDUCT.md). Also, check out some of our community resources including:

  - [Community Wiki](https://github.com/saltstack/community/wiki)
  - [Salt’s Contributor Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html)
  - [Join our Community Slack](https://via.vmw.com/salt-slack)
  - [IRC on LiberaChat](https://web.libera.chat/#salt)
  - [Salt Project YouTube channel](https://www.youtube.com/channel/UCpveTIucFx9ljGelW63-BWg)
  - [Salt Project Twitch channel](https://www.twitch.tv/saltprojectoss)

There are lots of ways to get involved in our community. Every month, there are around a dozen opportunities to meet with other contributors and the Salt Core team and collaborate in real time. The best way to keep track is by subscribing to the Salt Community Events Calendar.
If you have additional questions, email us at saltproject@vmware.com. We’re glad you’ve joined our community and look forward to doing awesome things with you!

**Comment by OrangeDog** [comment]
While I agree, it should be noted that there are already two solutions for your example:

```py
salt["pillar.get"]("myapp:cfg:pipeline:credentials:msk:keystore")
```
```py
pillar | traverse("myapp:cfg:pipeline:credentials:msk:keystore")
```

**Comment by hackery** [comment]
`traverse()` is nice, shame the [docs on Salt's custom filters](https://docs.saltproject.io/en/latest/topics/jinja/index.html#traverse) don't point out this behaviour.

**Comment by OrangeDog** [comment]
That documentation says exactly that...

**Comment by hackery** [comment]
> Traverse a dict or list using a colon-delimited target string. The target 'foo:bar:0' will return data['foo']['bar'][0] if this value exists, and will otherwise return the provided default value.

"if this value exists" is not explicit about intermediate targets.
